arch/x86/entry.S

```diff
@@ -32,70 +32,86 @@
 #include <usermode.h>
 #include <errno.h>
 
-.macro _handle_usermode cr3
+.macro MASK_USER_FLAGS
+    PUSHF
+    andl $~(USERMODE_FLAGS_MASK), (%_ASM_SP)
+    POPF
+.endm
+
+.macro _from_usermode switch_stack=0
+    SET_CR3 cr3
+    swapgs
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+.endm
+
+.macro _to_usermode switch_stack=0
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+    swapgs
+    SET_CR3 user_cr3
+.endm
+
+.macro cond_from_usermode
     testb $0x3, cpu_exc_cs(%_ASM_SP)
     jz 1f /* skip if from kernel mode */
-        SET_CR3 \cr3
-        swapgs
+        MASK_USER_FLAGS
+        _from_usermode
     1:
 .endm
 
-.macro enter_from_usermode
-    _handle_usermode cr3
+.macro cond_to_usermode
+    testb $0x3, cpu_exc_cs(%_ASM_SP)
+    jz 1f /* skip if from kernel mode */
+        _to_usermode
+    1:
 .endm
 
-.macro enter_to_usermode
-    _handle_usermode user_cr3
+.macro syscall_from_usermode switch_stack=1
+    _from_usermode switch_stack=\switch_stack
 .endm
 
-.macro syscall_from_usermode
-    SET_CR3 cr3
-    swapgs
-    SWITCH_STACK
-.endm
-
-.macro syscall_to_usermode
-    SWITCH_STACK
-    swapgs
-    SET_CR3 user_cr3
+.macro syscall_to_usermode switch_stack=1
+    _to_usermode switch_stack=\switch_stack
 .endm
 
 .macro exception_handler sym vec has_error_code
 ENTRY(entry_\sym)
-    enter_from_usermode
-
     .if \has_error_code == 0
         push $0
     .endif
-
-#if defined(__x86_64__)
-    movl $\vec, cpu_exc_vector(%_ASM_SP)
-#else
     push $\vec
-#endif
 
+    cond_from_usermode
     jmp handle_exception
 END_FUNC(entry_\sym)
 .endm
 
-.macro MASK_USER_FLAGS
-    PUSHF
-    andl $~(USERMODE_FLAGS_MASK), (%_ASM_SP)
-    POPF
+.macro check_syscall_exit
+    cmp $SYSCALL_EXIT, %_ASM_AX
+    jnz 1f
+        _from_usermode
+        jmp syscall_exit
+    1:
 .endm
 
-.macro interrupt_handler sym func
+.macro interrupt_handler sym func vec
 ENTRY(asm_interrupt_handler_\sym)
-    enter_from_usermode
-
-    MASK_USER_FLAGS
+    push $0 /* fake error_code */
+    push $\vec /* IRQ vector */
+    cond_from_usermode
 
     cld
```

Comment:
That shouldn't be needed any more. When we interrupted userland, `cond_from_usermode` sanitized the flags. If we interrupted kernel code, DF should be cleared as that's what C code requires anyway.

Same for [`handle_exception`](https://github.com/KernelTestFramework/ktf/pull/303/files#diff-b06dc21c831e6a8bbc838fc0fcfee4219ea48b7ad0e803fc15a3192bdebeaf58R120), the `CLD` can be dropped.

Reply:
Ok, will do.